core/src/main/scala/org/apache/spark/scheduler/MapStatus.scala

```diff
@@ -193,6 +198,11 @@ private[spark] object HighlyCompressedMapStatus {
     } else {
       0
     }
-    new HighlyCompressedMapStatus(loc, numNonEmptyBlocks, emptyBlocks, avgSize)
+    if (numNonEmptyBlocks < totalNumBlocks / 2) {
+      new HighlyCompressedMapStatus(loc, numNonEmptyBlocks, nonEmptyBlocks, avgSize, isSparse = true)
+    } else {
+      new HighlyCompressedMapStatus(loc, numNonEmptyBlocks, emptyBlocks, avgSize, isSparse = false)
```

Comment:
(a) `isSparse` is not the right name and (b) we might not even need to bother with tracking empty vs. non-empty, after we see the results of calling `runOptimize`.  In fact, if they are the same, I'd suggest we change this code to consistently track nonEmpty blocks, just for clarity.